File: specsy-core/src/main/scala/org/specsy/core/Path.scala
// Copyright © 2010-2012, Esko Luontola <www.orfjackal.net>
// This software is released under the Apache License 2.0.
// The license text is at http://www.apache.org/licenses/LICENSE-2.0

package org.specsy.core

import fi.jumi.api.drivers.TestId

object Path {
  def apply(indexes: Int*): Path = Path(TestId.of(indexes: _*))

  val Root: Path = Path()
}

case class Path(id: TestId) extends Ordered[Path] {

  def isRoot: Boolean = id.isRoot

  def parent: Path =
    if (id.isRoot) {
      this // XXX: fix usages so that they don't try to get root's parent
    } else {
      Path(id.getParent)
    }

  def isFirstChild: Boolean = id.isFirstChild

  def firstChild: Path = Path(id.getFirstChild)

  def nextSibling: Path = Path(id.getNextSibling)

  def isOn(that: Path): Boolean =
    this.id == that.id || this.id.isAncestorOf(that.id)

  def isBeyond(that: Path): Boolean = this.id.isDescendantOf(that.id)

  def compare(that: Path): Int = {
    this.id.compareTo(that.id)
  }
}

File: specsy-core/src/main/java/org/specsy/core/Path.java
// Copyright © 2010-2012, Esko Luontola <www.orfjackal.net>
// This software is released under the Apache License 2.0.
// The license text is at http://www.apache.org/licenses/LICENSE-2.0

package org.specsy.core;

import fi.jumi.api.drivers.TestId;

public class Path {

    public static final Path ROOT = new Path(TestId.ROOT);

    private final TestId id;

    private Path(TestId id) {
        this.id = id;
    }

    public static Path of(int... indexes) {
        return new Path(TestId.of(indexes));
    }

    public boolean isFirstChild() {
        return id.isFirstChild();
    }

    public Path firstChild() {
        return new Path(id.getFirstChild());
    }

    public Path nextSibling() {
        return new Path(id.getNextSibling());
    }

    public boolean isOn(Path that) {
        return this.id.equals(that.id) || this.id.isAncestorOf(that.id);
    }

    public boolean isBeyond(Path that) {
        return this.id.isDescendantOf(that.id);
    }

    public TestId toTestId() {
        return id;
    }

    @Override
    public String toString() {
        return "Path(" + id + ")";
    }
}
